### DISCUSSION TITLE
Making ai generated text to have markdown

### DISCUSSION THREAD

**Discussion by erbill** [steven-tey/novel#7] [Q&A]
I've tried to customize message content on route.ts to make it us markdown or HTML tags, it generates it but it doesn't visually applied. What kind of a workaround would you suggest to make text completions have related h1,h2,h3, <p> and bullet point tags as OpenAI generates it?

**Comment by steven-tey** [comment]
Great point – I can repro! This is probably something to do with the output and how its rendered via Tiptap...will look into it! Open to ideas & suggestions as well 👀 

![CleanShot 2023-06-17 at 23 18 19](https://github.com/steven-tey/novel/assets/28986134/a3d93210-0a8d-4f50-8a42-67a1c94d2e15)

**Reply by erbill** [reply]
Thanks! I looked into ui/editor folder to see if any changes made it update but had no success. 

How does OpenAI's chat page do it out of the box? Because one thing that comes to my mind is that when we manually type markdown or HTML tags in the editor it doesn't automatically render it as the slash-command box does perfectly. 

Is there any way to make the editor catch those manually typed tags and render them accordingly? 

If that's possible maybe the default char stream will be rendered with tags? Will let you know if I find anything ✌️

**Reply by steven-tey** [reply]
Do keep me posted if you find something - it'll be incredibly cool if we can support this 🤩

**Reply by steven-tey** [reply]
Update: I added the amazing https://github.com/aguingand/tiptap-markdown library today which is great for exporting Tiptap content as markdown. However, this still doesn't solve this issue because I think the library doesn't handle the streaming use case. 

- With streaming – links don't get transformed, numbered lists don't work: 

	![CleanShot 2023-06-27 at 22 14 50](https://github.com/steven-tey/novel/assets/28986134/735fa041-c290-4361-9907-dd837bed8ec4)

- Without streaming (`setContent` once the whole response is returned):

  	![CleanShot 2023-06-27 at 22 10 41](https://github.com/steven-tey/novel/assets/28986134/9dee460b-c23c-42ad-8b15-b722b8fdaead)

cc'ing the creator @aguingand who might be able to point is in the right direction here 👀

**Reply by aguingand** [reply]
Maintainer of this package here. I've already seen some questions concerning streaming and I answered that it may be difficult because of the flatness of markdown. Nevertheless, a PR or a well detailed issue (with concrete code example + input/expected output) is welcomed.

**Reply by steven-tey** [reply]
@aguingand really appreciate you weighing in here! Will come up with a PR / issue! :)

**Reply by Morriz** [reply]
The way I (and others) do it (using `ReactMarkdown`) is to re-render the MD element on each streaming token update. No perf issues as it's just html rendering...

**Reply by RicSala** [reply]
@Morriz can we see how?

**Reply by Morriz** [reply]
Just update the content to render with the new token and react will render because of the prop update. Basic react stuff actually

**Reply by gediminastub** [reply]
Could You share a short code snippet for that? Would be very grateful! @Morriz

**Reply by Gyurmatag** [reply]
Yeah, @Morriz can you please share it with us?

**Reply by Morriz** [reply]
Geez, please go do some basic react tutorial and pay attention to the render function...

**Reply by gediminastub** [reply]
@Morriz You do this inside of TipTap?

**Reply by levi-katarok** [reply]
Just curious how you actually interact with the setContent, I don't see that as a prop on the react component. Any help is appreciated !

**Reply by Morriz** [reply]
I will share the file content in question, but please figure it out from there yourself and stop asking me how to do 1+1

```ts
import styled from '@emotion/styled'
import { Button } from '@mantine/core'
import { useMemo } from 'react'
import { FormattedMessage } from 'react-intl'
import ReactMarkdown from 'react-markdown'
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter'
import { vscDarkPlus } from 'react-syntax-highlighter/dist/esm/styles/prism'
import rehypeKatex from 'rehype-katex'
import remarkGfm from 'remark-gfm'
import remarkMath from 'remark-math'
import { CopyButton } from './copy-button'

const Code = styled.div`
  padding: 0;
  border-radius: 0.25rem;
  overflow: hidden;

  & > div {
    margin: 0 !important;
  }

  .fa {
    font-style: normal !important;
  }
`

const Header = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  // background: #191919;
  height: 2.5rem;
  padding: 0.1rem 0.1rem 0 0.5rem;

  .mantine-Button-label {
    display: flex;
    align-items: center;

    * {
      font-size: 90%;
    }
  }
`

const ImagePreview = styled.div`
  text-align: center;

  img {
    max-width: 30rem !important;
    display: block;
  }
`

export interface MarkdownProps {
  content: string
  className?: string
}

export function Markdown(props: MarkdownProps) {
  const classes = useMemo(() => {
    const classes: string[] = ['prose', 'dark:prose-invert']

    if (props.className) {
      classes.push(props.className)
    }

    return classes
  }, [props.className])

  const elem = useMemo(
    () => (
      <div className={classes.join(' ')}>
        <ReactMarkdown
          remarkPlugins={[remarkGfm, remarkMath]}
          rehypePlugins={[rehypeKatex]}
          linkTarget="_blank"
          components={{
            ol({ start, children }) {
              return (
                <ol start={start ?? 1} style={{ counterReset: `list-item ${start || 1}` }}>
                  {children}
                </ol>
              )
            },
            code({ inline, className, children, ...props }) {
              const match = /language-(\w+)/.exec(className || '')
              const code = String(children)
              return !inline ? (
                <>
                  <Code>
                    <Header>
                      {code.startsWith('<svg') && code.includes('</svg>') && (
                        <Button
                          variant="subtle"
                          size="sm"
                          compact
                          onClick={() => {
                            const blob = new Blob([code], { type: 'image/svg+xml' })
                            const url = URL.createObjectURL(blob)
                            const a = document.createElement('a')
                            a.href = url
                            a.download = 'image.svg'
                            a.click()
                          }}
                        >
                          <i className="fa fa-download" />
                          <span>
                            <FormattedMessage defaultMessage="Download SVG" />
                          </span>
                        </Button>
                      )}
                      <CopyButton value={code} />
                    </Header>
                    <SyntaxHighlighter
                      children={code}
                      style={vscDarkPlus as any}
                      language={match?.[1] || 'text'}
                      PreTag="div"
                      {...props}
                    />
                  </Code>
                  {code.startsWith('<svg') && code.includes('</svg>') && (
                    <ImagePreview>
                      <img alt="code" src={`data:image/svg+xml;base64,${btoa(code)}`} />
                    </ImagePreview>
                  )}
                </>
              ) : (
                <code className={className} {...props}>
                  {children}
                </code>
              )
            },
          }}
        >
          {props.content}
        </ReactMarkdown>
      </div>
    ),
    [props.content, classes],
  )

  return elem
}
```

**Comment by SamikshyaKarki0** [comment]
I am looking for AI job as well

**Comment by SamikshyaKarki0** [comment]
Yeah they will

**Comment by andrewdoro** [comment]
This PR adds a guide for having AI features like Notion, it also works with Markdown streaming.
https://github.com/steven-tey/novel/pull/297
[live example](https://novel-r5wzv908o-elegance.vercel.app/)
